INTAP-85: Issue entity, bundle structure.

Florin1 opened the pull request.

Commit:
INTAP-85: Issue entity, bundle structure.

Commit:
INTAP-85: Issue Entity
- fixed missing new line

Commit:
INTAP-85: Issue entity, added children and parent relationship.

Commit:
INTAP-86: Migrations and Installer
- added v1_0 migration : init db schema , issue enums , issue note, issue activity
- issue entity fix , collaborators db rename
- issue entity , added parent and children fields

Commit:
INTAP-87: Issue enums data fixtures
- type
- resolution
- priority

Commit:
INTAP-88: Issue Form Type.
- Issue form type
- Issue form handler
- Issue form event subscriber
- Issue validation
- Issue entity/repo update

Commit:
INTAP-90: Issue UI CRUD
- Issue Controller
- Issue Views
- moved routes config

Commit:
INTAP-98: Grid issues
- added grid config
- added translations

Commit:
INTAP-91 : Issue Search
- added search config
- added search result template

Commit:
INTAP-89: Issue API CRUD
- added config for api crud

Commit:
INTAP-92: Import/Export
- import/export config
- issue template fixture
- issue entity update for export

Commit:
INTAP-94: Issue workflow
- added migration for status enum on issue
- workflow configuration
- fixture updates for status field on issue

Commit:
INTAP-95: Update Collaborators
- process listener on issue note creation
- collaborators population

Commit:
INTAP-105: Issue Update Process
- adding a note on issue updated field

Commit:
INTAP-99: Grid Issues Sort
- issue sort in priority by value
- grid columns update for enum fields
- migration update for enum fields, changed owner to system
- issue index twig update

Commit:
INTAP-101: Dashboard issues grid
- added dashboard fixtures for new widget
- added recent issue widget template
- translations
- dashboard widget config
- datagrid for recent issues

Commit:
INTAP-100: Dashboard issues chart
- added dashboard chart widget
- added chart action in issue controller
- added chart data provider
- updated datagrig fixtures with chart widget

Commit:
INTAP-102: Grid User View Page
- added config for new grid on user view page
- added additional data placeholder, placeholder filter
- replaced doctrine manager with doctrine helper in services declaration

Commit:
INTAP-103: User View page Create Issue
- issue form handler update
- issue update view update
- user view page update
- installer update

Commit:
INTAP-106: Add subtask to story
- issue form handler update
- issue form type update
- issue view page update

Commit:
INTAP-104: Navigation
- added navigation menu

AlexPuscas commented on src/Academic/Bundle/BugTrackingBundle/Controller/IssueController.php:
```diff
@@ -0,0 +1,120 @@
+
+namespace Academic\Bundle\BugTrackingBundle\Controller;
+
+use Academic\Bundle\BugTrackingBundle\Form\Handler\IssueHandler;
```
Please reorder the dependencies.

``` php
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Template;

use Symfony\Bundle\FrameworkBundle\Controller\Controller;
use Symfony\Component\HttpFoundation\RedirectResponse;

use Academic\Bundle\BugTrackingBundle\Entity\Issue;
use Academic\Bundle\BugTrackingBundle\Form\Handler\IssueHandler; 
use Academic\Bundle\BugTrackingBundle\Form\Type\IssueType;
```

AlexPuscas commented on src/Academic/Bundle/BugTrackingBundle/Controller/IssueController.php:
```diff
@@ -0,0 +1,120 @@
+     * )
+     * @Template
+     */
+    public function indexAction()
```
All methods with parameters should have @param docblock.
All methods that return something should have @return docblock.

AlexPuscas commented on src/Academic/Bundle/BugTrackingBundle/Controller/IssueController.php:
```diff
@@ -0,0 +1,120 @@
+     */
+    public function viewAction(Issue $issue)
+    {
+        return [
```
This return can be compressed to one line.

AlexPuscas commented on src/Academic/Bundle/BugTrackingBundle/DependencyInjection/Configuration.php:
```diff
@@ -0,0 +1,24 @@
+        $treeBuilder = new TreeBuilder();
+        $rootNode = $treeBuilder->root('academic_bug_tracking');
+
+        // Here you should define the parameters that are allowed to
```
Is this comment required?

Commit:
INTAP-104: Navigation
- page title updates

AlexPuscas commented on src/Academic/Bundle/BugTrackingBundle/Resources/config/importexport.yml:
```diff
@@ -0,0 +1,38 @@
+     academic.bug_tracking.importexport.template_fixture.issue:
+         class: '%academic.bug_tracking.importexport.template_fixture.issue.class%'
+         tags:
+             - { name: oro_importexport.template_fixture }
```
Please add a new line at the end of all files, even the .yml ones.

Commit:
Pull reuqest fixes:
- reordered dependencies
- yml new line at end of file
- removed comments
- view/update action button update

dandries commented on src/Academic/Bundle/BugTrackingBundle/Form/EventListener/IssueSubscriber.php:
```diff
@@ -0,0 +1,84 @@
+
+use Doctrine\ORM\EntityRepository;
+
+use Oro\Bundle\UserBundle\Entity\User;
```
Separate different vendors.

dandries commented on src/Academic/Bundle/BugTrackingBundle/Migrations/Data/ORM/LoadIssueData.php:
```diff
@@ -0,0 +1,160 @@
+        ],
+    ];
+
+    public function load(ObjectManager $manager)
```
Add phpdoc.

dandries commented on src/Academic/Bundle/BugTrackingBundle/Migrations/Schema/v1_1/AcademicBugTrackingBundle.php:
```diff
@@ -0,0 +1,56 @@
+namespace Academic\Bundle\BugTrackingBundle\Migrations\Schema\v1_1;
+
+use Doctrine\DBAL\Schema\Schema;
+use Oro\Bundle\MigrationBundle\Migration\Migration;
```
Separate vendors.

dandries commented on src/Academic/Bundle/BugTrackingBundle/Migrations/Schema/v1_1/AcademicBugTrackingBundle.php:
```diff
@@ -0,0 +1,56 @@
+use Oro\Bundle\EntityExtendBundle\Migration\Extension\ExtendExtension;
+use Oro\Bundle\EntityExtendBundle\Migration\Extension\ExtendExtensionAwareInterface;
+use Oro\Bundle\EntityExtendBundle\EntityConfig\ExtendScope;
+use Academic\Bundle\BugTrackingBundle\Entity\Issue;
```
Separate vendors.

dandries commented on src/Academic/Bundle/BugTrackingBundle/Resources/translations/messages.en.yml:
```diff
@@ -0,0 +1,64 @@
+                view:
+                    label: 'View'
+                edit:
+                    label: 'Edit'
```
Add new line.

dandries commented on src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/index.html.twig:
```diff
@@ -0,0 +1,28 @@
+            'entity_label': 'academic.bugtracking.issue.entity_label'|trans
+        }) }}
+    </div>
+{% endblock %}
```
Add new line.

dandries commented on src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/searchResult.html.twig:
```diff
@@ -0,0 +1,16 @@
+{'title': 'academic.bugtracking.issue.resolution.label'|trans, 'value': entity.resolution ? entity.resolution : 'N/A'|trans},
+{'title': 'academic.bugtracking.issue.reporter.label'|trans, 'value': entity.reporter ? entity.reporter : 'N/A'|trans},
+{'title': 'academic.bugtracking.issue.assignee.label'|trans, 'value': entity.assignee ? entity.assignee : 'N/A'|trans},
+] %}
```
Add new line.

dandries commented on src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/update.html.twig:
```diff
@@ -0,0 +1,100 @@
+    } %}
+
+    {{ parent() }}
+{% endblock content_data %}
```
Add new line.

dandries commented on src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/view.html.twig:
```diff
@@ -0,0 +1,94 @@
+    {% set data = {'dataBlocks': dataBlocks} %}
+
+    {{ parent() }}
+{% endblock content_data %}
```
Add new line.

Commit:
Pull request fixes #2

Commit:
INTAP-93: Fixtures Demo Data

Commit:
INTAP-107:Functional tests

Commit:
INTAP-108: Unit tests

AlexPuscas commented:
Make sure you can only create a Subtask to a Story

AlexPuscas commented:
Do you really need DependencyInjection\Configuration?

AlexPuscas commented on src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Controller/IssueControllerTest.php:
```diff
@@ -0,0 +1,99 @@
+        $this->assertContains("Successfully saved issue", $crawler->html());
+    }
+
+    public function testUpdateAction()
```
You can pass the ID of the issue from one test to another, and so perform all CRUD actions for the same issue, by returning the ID and depending on the previous test.

AlexPuscas commented on src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Controller/IssueControllerTest.php:
```diff
@@ -0,0 +1,99 @@
+        $result = $this->client->getResponse();
+
+        $this->assertHtmlResponseStatusCodeEquals($result, 200);
+        $this->assertContains('General Information', $result->getContent());
```
You should test something specific for an issue, like name or code.

AlexPuscas commented on src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Entity/Repository/IssueRepositoryTest.php:
```diff
@@ -0,0 +1,34 @@
+    {
+        $result = reset($this->repository->getIssuesGroupedByStatus());
+
+        $this->assertArrayHasKey('status', $result);
```
You should also check the result of the query.

yshyshkin replied:
Please, provide better assertion of method results.

Commit:
Removed api delete issue.

AlexPuscas commented on src/Academic/Bundle/BugTrackingBundle/Tests/Unit/Form/Type/IssueTypeTest.php:
```diff
@@ -0,0 +1,104 @@
+
+    public function testBuildForm()
+    {
+        $this->issueType->buildForm($this->builder, []);
```
You should actually test the submit of the form. See http://symfony.com/doc/current/form/unit_testing.html

AlexPuscas commented:
Please unit test classes from DependencyInjection folder.

Commit:
PR fixes.

Commit:
Datagid fix.

yshyshkin commented on src/Academic/Bundle/BugTrackingBundle/Form/Type/IssueType.php:
```diff
@@ -0,0 +1,129 @@
+        } else {
+            $builder->add(
+                'parent',
+                EntityType::class, array(
```
Please, maintain consistency and use short syntax array all over the code.

yshyshkin commented:
Please, fix code style issues:

```
maniac@laphere:/opt/www/academ-fo$ phpcs --encoding=utf-8 --extensions=php --standard=psr2 src/ -p
..........EE....E..............


FILE: ...fo/src/Academic/Bundle/BugTrackingBundle/Form/Type/IssueType.php
----------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 84 | ERROR | [x] Only one argument is allowed per line in a
    |       |     multi-line function call
 87 | ERROR | [x] Closing parenthesis of a multi-line function call
    |       |     must be on a line by itself
----------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: .../Bundle/BugTrackingBundle/Form/EventListener/IssueSubscriber.php
----------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 60 | ERROR | [x] Expected 0 spaces before closing bracket; newline
    |       |     found
 66 | ERROR | [x] Expected 0 spaces before closing bracket; newline
    |       |     found
----------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: ...ackingBundle/Tests/Functional/Controller/IssueControllerTest.php
----------------------------------------------------------------------
FOUND 3 ERRORS AFFECTING 1 LINE
----------------------------------------------------------------------
 78 | ERROR | [x] Multi-line function call not indented correctly;
    |       |     expected 8 spaces but found 12
 78 | ERROR | [x] Only one argument is allowed per line in a
    |       |     multi-line function call
 78 | ERROR | [x] Closing parenthesis of a multi-line function call
    |       |     must be on a line by itself
----------------------------------------------------------------------
PHPCBF CAN FIX THE 3 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```

yshyshkin commented on src/Academic/Bundle/BugTrackingBundle/Form/EventListener/IssueSubscriber.php:
```diff
@@ -0,0 +1,85 @@
+    protected function getUsersClosure()
+    {
+        return function (EntityRepository $entityRepository) {
+            return $entityRepository->createQueryBuilder('user')
```
Please, move direct query builder manipulation to entity repository and then call appropriate method here- form layer should not be aware about database structure.

yshyshkin commented on src/Academic/Bundle/BugTrackingBundle/Placeholder/PlaceholderFilter.php:
```diff
@@ -0,0 +1,27 @@
+     */
+    protected $request;
+
+    public function __construct(Request $request)
```
Please, add missing phpdoc.

yshyshkin commented on src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/index.html.twig:
```diff
@@ -0,0 +1,28 @@
+    entity_class: entity_class,
+    exportProcessor: 'academic.bug_tracking_issue',
+    exportTemplateProcessor: 'academic.bug_tracking_issue',
+    exportTitle: 'Export',
```
Please, translate this label.

yshyshkin commented on src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/index.html.twig:
```diff
@@ -0,0 +1,28 @@
+    exportTemplateProcessor: 'academic.bug_tracking_issue',
+    exportTitle: 'Export',
+    importProcessor: 'academic.bug_tracking_issue',
+    importTitle: 'Import',
```
Please, translate this label.

Commit:
PR fixes.

yshyshkin commented on src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Entity/Repository/IssueRepositoryTest.php:
```diff
@@ -0,0 +1,36 @@
+        $result = reset($this->repository->getIssuesGroupedByStatus());
+
+        $this->assertArrayHasKey('status', $result);
+        $this->assertEquals($result['status'], 'open');
```
Please, don't forget that first argument should contain expected value, and second argument - actual value.

yshyshkin commented on src/Academic/Bundle/BugTrackingBundle/Placeholder/PlaceholderFilter.php:
```diff
@@ -0,0 +1,31 @@
+    protected $request;
+
+    /**
+     * PlaceholderFilter constructor.
```
Please, avoid useless (e.g. autogenerated) comments. Comments should provide valuable information for developer, if they don't do that - it's better to remove them. In this case would be better to have only phpdoc in method comment.
